### DISCUSSION TITLE
Forge script failed with `project id required in the url`

### DISCUSSION THREAD

**Discussion by worca333** [Cyfrin/foundry-full-course-cu#1898] [Q&A]
Hello fellas!

I have been learning foundry fundamental about how to compile and deploy contracts and this time I tried with a testnet (sepolia).
What I found weird is that I have to create a custom node in alchemy for an RPC url.

Then my question goes **why can't we just use the one from Metamask?**

Below are the commands and env values I used: (Don't worry, the private key is for test net. I AM SAFE :))
```
PRIVATE_KEY=3b2eb22e160a33f6d0377f22d12fefbd6ee1881018d7a08ea64322ed819f5411
SEPOLIA_RPC_URL=https://sepolia.infura.io/v3/

forge script script/DeploySimpleStorage.s.sol --rpc-url $SEPOLIA_RPC_URL --private-key $PRIVATE_KEY --broadcast
```

**Comment by EngrPips** [comment]
Hello @worca333, Honestly, I don't think that is weird. If you want to send transactions to the blockchain directly, you should have a node to which you are connected. If we can all use  Metamask node, don't you think we will overload it with requests?

**Reply by worca333** [reply]
Thanks for your answer, @EngrPips 
That's interesting. So that means Metamask is (or contains) a blockchain node to which we can send transactions for the blockchain network. I thought it was nothing but a DApp that communicates with the existing networks just like I did in CLI.

In that case, can I assume all DApps would come with their own network nodes?

**Reply by EngrPips** [reply]
<img width="1440" alt="Screenshot 2024-06-28 at 15 15 59" src="https://github.com/Cyfrin/foundry-full-course-cu/assets/95760918/ffd1f56f-b240-4021-a17c-a789a0c366ec">

Whenever I am on the fuji Avalanche chain in my metamask and I send a transaction, That `rpc-url` is the node to which my transaction would be sent to and yes, every dapp needs to be connected to a blockchain node to read and write to the blockchain

**Reply by worca333** [reply]
Absolutely.
We need a transaction sender and receiver (blockchain node) who may put my transaction into the mempool.
And what Metamask does is to communicate with the rpc-url the same way as I do on local. Hence I got the error saying **project id is required**.
So the problem is that Metamask is a registered app with the specific network while my CLI is not. That's why I should create a DApp in alchemy or something and call through that. Did I get it right?

**Reply by EngrPips** [reply]
This makes sense to me.
